src/Compilers/CSharp/Portable/Binder/Binder_Deconstruct.cs

```diff
@@ -282,21 +290,59 @@ public DeconstructionVariable(ArrayBuilder<DeconstructionVariable> variables, CS
         }
 
         /// <summary>
-        /// For cases where the RHS of a deconstruction-assignment has no type (TupleLiteral), we squint and look at the LHS as a tuple type to give the RHS a type.
+        /// For cases where the RHS of a deconstruction-declaration is a tuple literal, we merge type information from both the LHS and RHS.
+        /// For cases where the RHS of a deconstruction-assignment is a tuple literal, the type information from the LHS determines the merged type, since all variables have a type.
+        /// Returns null if a merged tuple type could not be fabricated.
         /// </summary>
-        private static TypeSymbol MakeTupleTypeFromDeconstructionLHS(ArrayBuilder<DeconstructionVariable> topLevelCheckedVariables, DiagnosticBag diagnostics, CSharpCompilation compilation)
+        private static TypeSymbol MakeMergedTupleType(ArrayBuilder<DeconstructionVariable> lhsVariables, BoundTupleLiteral rhsLiteral, CSharpSyntaxNode syntax, DiagnosticBag diagnostics, CSharpCompilation compilation)
         {
-            var typesBuilder = ArrayBuilder<TypeSymbol>.GetInstance(topLevelCheckedVariables.Count);
-            foreach (var variable in topLevelCheckedVariables)
+            int leftLength = lhsVariables.Count;
+            int rightLength = rhsLiteral.Arguments.Length;
+
+            var typesBuilder = ArrayBuilder<TypeSymbol>.GetInstance(lhsVariables.Count);
+            for (int i = 0; i < rightLength; i++)
             {
-                if (variable.HasNestedVariables)
-                {
-                    typesBuilder.Add(MakeTupleTypeFromDeconstructionLHS(variable.NestedVariables, diagnostics, compilation));
-                }
-                else
+                BoundExpression element = rhsLiteral.Arguments[i];
+                TypeSymbol mergedType = element.Type;
+
+                if (i < leftLength)
```

Comment:
Below is the test for that scenario. The merged type is `(string, string, int)`, so we have an error reported for cardinality mismatch in conversion.
I also tried with the RHS being `(null, "hello", null)` and the error is the same. I'll add that test.


```C#

        [Fact]
        public void TypeMergingWithTooManyRightElements()
        {
            string source = @"
class C
{
    static void Main()
    {
        (string x1, var x2) = (null, ""hello"", 3);
    }
}
";
            var comp = CreateCompilationWithMscorlib(source, references: new[] { ValueTupleRef, SystemRuntimeFacadeRef });
            comp.VerifyDiagnostics(
                // (6,9): error CS8211: Cannot deconstruct a tuple of '3' elements into '2' variables.
                //         (string x1, var x2) = (null, "hello", 3);
                Diagnostic(ErrorCode.ERR_DeconstructWrongCardinality, @"(string x1, var x2) = (null, ""hello"", 3);").WithArguments("3", "2").WithLocation(6, 9)
                );
        }
```